web/src/engine/predictive-text/worker-thread/src/main/correction/distance-modeler.ts

```diff
@@ -157,68 +258,232 @@ export class SearchNode {
+    let keySet: Set<string> = new Set();
+
+    const subsetMap = subsetByChar(startSubset, subIndex, this.toKey);
+    for(const [char, subset] of subsetMap.entries()) {
+      // build new node for the next char.
+      let calculation = this.calculation;
+      if(char) {
+        const childTraversal = traversal.child(char);
+        // These cases - where there's no match in the lexicon - are bundled
+        // after this for-loop.
+        //
+        // ... except for when there ARE no children at all.  But, for those
+        // cases... a substitution would be invalid - we can't substitute for
+        // each char in the insert string, so abort.  It's better matched by an
+        // 'insert' edge or a by different input from a sibling subset.
+        if(!childTraversal) {
+          continue;
+        }
+
+        calculation = calculation.addInputChar({key: char});
+        calculation = calculation.addMatchChar({key: char, traversal: childTraversal});
+      } // else we COULD bundle these as a single subset... but it's likely not that important.
+      // after all, the cost ISN'T shifting, so we'll process 'em almost immediately and move on.
+
+      keySet.add(char);
+
+      const node = new SearchNode(this);
+      node.calculation = calculation;
+      node.currentTraversal = calculation.lastMatchEntry?.traversal ?? this.root;
+      node.partialEdge.subsetSubindex++;
+      // Append the newly-processed char to the subset's `insert` string.
+      node.partialEdge.transformSubset = {...subset, key: startSubset.key, insert: subset.insert + char};
+      nodesToReturn.push(node);
+    };
+
+    // These come at a notably higher cost (due to required edit) and are less likely to be processed.
+    // It's best to batch them so we only need a single in-memory node to represent them, rather than
+    // one node per transform / subset, which can scale very rapidly.
+    let calculation = this.calculation.addInputChar({ key: SENTINEL_CODE_UNIT });
+
+    for(const child of traversal.children()) {
+      let childCalc =  calculation.addMatchChar({key: child.char, traversal: child.traversal()});
+      let childSubset: TransformSubset<number>;
+
+      if(keySet.has(child.char)) {
+        // OK, so we built a set that successfully matched this char.  Other paths that
+        // emit a char after keying don't match it - we want to combine that path here.
+        // If there are multiple 'insert' chars, missing on one shouldn't force a 'miss'
+        // on the other subpaths that may follow this with valid char matches.
+        const nonMatchSubsets = [...subsetMap.values()].filter(c => {
+          // Ignore entries that 'key out' - there's no char left to act as a substitute.
+          // Also ignore the entry that actually did match - it's a match, not substitute.
+          return c.key != child.char && c.key != '';
+        });
+        childSubset = mergeSubset(nonMatchSubsets, startSubset.key);
```

Comment:
Noting this as the answer to https://github.com/keymanapp/keyman/pull/14581/files#r2301097783.